src/mame/drivers/klax.cpp

```diff
@@ -217,34 +236,89 @@ void klax_state::klax(machine_config &config)
+	map(0x2800, 0x28ff).w(FUNC(klax_bootleg_state::audio_sample_w));
+	map(0x3000, 0x3000).w(FUNC(klax_bootleg_state::audio_ctrl_w));
+	map(0x3800, 0x3800).nopw(); // ?
+	map(0x4f80, 0x4fff).ram();
+	map(0x8000, 0xffff).bankr("rombank");
 }
 
-void klax_state::klax5bl(machine_config &config)
+void klax_bootleg_state::klax5bl(machine_config &config)
 {
-	klax(config);
+	klax_base(config);
 
-	m_maincpu->set_addrmap(AS_PROGRAM, &klax_state::klax5bl_map);
+	m_maincpu->set_addrmap(AS_PROGRAM, &klax_bootleg_state::klax5bl_map);
 
-	config.device_remove("oki"); // no 6295 here
+	Z80(config, m_audiocpu, 14.318181_MHz_XTAL / 2); /* ? */
+	m_audiocpu->set_addrmap(AS_PROGRAM, &klax_bootleg_state::bootleg_sound_map);
 
-	z80_device &audiocpu(Z80(config, "audiocpu", 6000000)); /* ? */
-	audiocpu.set_addrmap(AS_PROGRAM, &klax_state::bootleg_sound_map);
+	MCFG_MACHINE_START_OVERRIDE(klax_bootleg_state, klax5bl)
 
+	m_mob->set_origin(8, 7); // sprites are offset in bootlegs
 	m_gfxdecode->set_info(gfx_klax5bl);
 
-	// guess, probably something like this
-	// 2 x msm at least on bootleg set 2 (ic18 and ic19)
-	MSM5205(config, "msm", 375000);     /* ? */
-//  msm.vck_legacy_callback().set(FUNC(klax_state::m5205_int1));    /* interrupt function */
-//  msm.set_prescaler_selector(msm5205_device::MSM5205_S96_4B);     /* 4KHz 4-bit */
-//  msm.add_route(ALL_OUTPUTS, "mono", 0.25);
+	MSM5205(config, m_msm[0], 384000);
+	m_msm[0]->vck_callback().set(FUNC(klax_bootleg_state::m5205_int1));
+	m_msm[0]->set_prescaler_selector(msm5205_device::S64_4B); /* ? */
+	m_msm[0]->add_route(ALL_OUTPUTS, "mono", 1.00);
+
+	MSM5205(config, m_msm[1], 384000);
+	m_msm[1]->set_prescaler_selector(msm5205_device::S64_4B);
+	m_msm[1]->add_route(ALL_OUTPUTS, "mono", 1.00);
+}
+
+uint8_t klax_bootleg_state::audio_ram_r(offs_t offset)
+{
+	return m_audio_ram[offset ^ 1];
+}
+
+void klax_bootleg_state::audio_ram_w(offs_t offset, uint8_t data)
+{
+	m_audio_ram[offset ^ 1] = data;
+}
+
+void klax_bootleg_state::audio_sample_w(offs_t offset, uint8_t data)
+{
+	if (offset & 2) m_audio_sample[0] = data;
+	if (offset & 1) m_audio_sample[1] = data;
+}
+
+void klax_bootleg_state::audio_ctrl_w(uint8_t data)
+{
+	m_rombank->set_entry(data & 3);
+
+	m_msm[0]->reset_w(data & 4);
+	m_msm[1]->reset_w(data & 8);
+
+	static const float gain[4] = { 1.0, 0.5, 0.75, 0.25 };
+	m_msm[0]->set_output_gain(ALL_OUTPUTS, gain[(data & 0x30) >> 4]);
+	m_msm[1]->set_output_gain(ALL_OUTPUTS, gain[(data & 0xc0) >> 6]);
+}
+
+void klax_bootleg_state::init_klax5bl()
+{
+	// RAM shared between main and audio CPU
+	m_audio_ram = std::make_unique<uint8_t[]>(0x80);
+	m_maincpu->space(AS_PROGRAM).install_ram(0x3f0f00, 0x3f0f7f, m_audio_ram.get());
+	m_audiocpu->space(AS_PROGRAM).install_read_handler(0x4f00, 0x4f7f, read8sm_delegate(*this, FUNC(klax_bootleg_state::audio_ram_r)));
+	m_audiocpu->space(AS_PROGRAM).install_write_handler(0x4f00, 0x4f7f, write8sm_delegate(*this, FUNC(klax_bootleg_state::audio_ram_w)));
```

Comment:
Please do this in the appropriate memory maps.  You can use `.share(...)` and a `required_shared_ptr` member.

Reply:
Is there actually a way to do this when the main and audio CPUs use two different data bus widths? I got the impression from e.g. the megcd driver that read/write delegates were needed for that reason.

Reply:
You still need read/write handlers to access 8-bit RAM from a 16-bit address space (or even worse, 16-bit peripherals from an 8-bit address space, see the Dooyong text layer).  Use `.ram().share(...)` in the 8-bit address space.

My objection is to adding an init function for something that should be done in address maps.

Reply:
Alright, that makes sense. Should be better/cleaner now.